### PR TITLE
[Saved Object Service] Adds Repository Factory Provider

### PR DESCRIPTION
**Pull request by bandinib-amzn**
### Description

The **Repository Factory Provider** in OpenSearch-Dashboards will be responsible for creating and managing instances of the Repository (e.g. SavedObjectRepository, PostgresRepository, DynamoDBRepository etc.), which is used to interact with the metadata storage that stores the saved objects. Currently we have an repository interface named [ISavedObjectRepository](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/repository.ts#L134), and the [SavedObjectRepository](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/repository.ts#L139) is the implementation, which use an OpenSearch index as its data store. This approach would make the implementation of [ISavedObjectRepository](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/repository.ts#L134) replaceable by plugin.

More details can be found in [this](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3954) PR.

### Check List

- [X] All tests pass
  - [X] `yarn test:jest`
  - [X] `yarn test:jest_integration`
  - [X] `yarn test:ftr`
- [X] New functionality includes testing.
- [X] New functionality has been documented.
- [X] Update [CHANGELOG.md](./../CHANGELOG.md)
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by AMoo-Miki — APPROVED** [review]

File: src/core/server/legacy/legacy_service.ts
Comment (zengyan-amazon) [inline src/core/server/legacy/legacy_service.ts:278]: lets name this function as `setRepositoryFactoryProvider` to be consistent as setting client factory provider function

Comment (bandinib-amzn) [reply src/core/server/legacy/legacy_service.ts]: Changed the function name.

**Review by zengyan-amazon — APPROVED** [review]

File: src/core/server/saved_objects/service/lib/scoped_client_provider.ts
Comment (zhongnansu) [inline src/core/server/saved_objects/service/lib/scoped_client_provider.ts:51]: when is this param needed? If not, can we remove?

File: src/core/server/saved_objects/saved_objects_service.ts
Comment (Flyingliuhub) [inline src/core/server/saved_objects/saved_objects_service.ts:367]: `set` to `registered`?

**Review by Flyingliuhub — DISMISSED** [review]

File: src/core/server/saved_objects/saved_objects_service.test.ts
Comment (Flyingliuhub) [inline src/core/server/saved_objects/saved_objects_service.test.ts:174]: I saw cypress test failed, I'm assuming that the failed cypress test should not related to your PR. maybe double check before merge in. Thanks

File: src/core/server/saved_objects/saved_objects_service.ts
Comment (bandinib-amzn) [reply src/core/server/saved_objects/saved_objects_service.ts:367]: I changed the function name to `setRepositoryFactoryProvider` to be consistent as setting client factory provider function.

File: src/core/server/saved_objects/saved_objects_service.test.ts
Comment (bandinib-amzn) [reply src/core/server/saved_objects/saved_objects_service.test.ts:174]: The failed cypress test is not related to this PR.

**Review by Flyingliuhub — APPROVED** [review]

File: src/core/server/saved_objects/service/lib/scoped_client_provider.ts
Comment (bandinib-amzn) [reply src/core/server/saved_objects/service/lib/scoped_client_provider.ts]: Good catch. Thanks! I removed as we don't see use case for now for `OpenSearchClient`. If in future, if we come up with use case where `OpenSearchClient` needed, we can add back as optional parameter.

**Review by zhongnansu — APPROVED** [review]
ship it!

**Review by zengyan-amazon — APPROVED** [review]

**Comment by zhongnansu** [comment]
will merge after CI completes

**Pull request merged**
